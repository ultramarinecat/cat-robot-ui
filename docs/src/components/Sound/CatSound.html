<!DOCTYPE html>

<html>
<head>
  <title>CatSound.js</title>
  <meta http-equiv="content-type" content="text/html; charset=UTF-8">
  <meta name="viewport" content="width=device-width, target-densitydpi=160dpi, initial-scale=1.0; maximum-scale=1.0; user-scalable=0;">
  <link rel="stylesheet" media="all" href="../../../docco.css" />
</head>
<body>
  <div id="container">
    <div id="background"></div>
    
      <ul id="jump_to">
        <li>
          <a class="large" href="javascript:void(0);">Jump To &hellip;</a>
          <a class="small" href="javascript:void(0);">+</a>
          <div id="jump_wrapper">
          <div id="jump_page_wrapper">
            <div id="jump_page">
              
                
                <a class="source" href="../App/App.html">
                  src/components/App/App.js
                </a>
              
                
                <a class="source" href="../App/ErrorBoundary.html">
                  src/components/App/ErrorBoundary.js
                </a>
              
                
                <a class="source" href="../App/i18n.html">
                  src/components/App/i18n.js
                </a>
              
                
                <a class="source" href="../Auth/Auth.html">
                  src/components/Auth/Auth.js
                </a>
              
                
                <a class="source" href="../Auth/AuthActions.html">
                  src/components/Auth/AuthActions.js
                </a>
              
                
                <a class="source" href="../Auth/AuthReducer.html">
                  src/components/Auth/AuthReducer.js
                </a>
              
                
                <a class="source" href="../Button/Button.html">
                  src/components/Button/Button.js
                </a>
              
                
                <a class="source" href="../Cat/Cat.html">
                  src/components/Cat/Cat.js
                </a>
              
                
                <a class="source" href="../Cat/CatActions.html">
                  src/components/Cat/CatActions.js
                </a>
              
                
                <a class="source" href="../Cat/CatReducer.html">
                  src/components/Cat/CatReducer.js
                </a>
              
                
                <a class="source" href="../Cat/fulltilt.html">
                  src/components/Cat/fulltilt.js
                </a>
              
                
                <a class="source" href="../Connection/Connection.html">
                  src/components/Connection/Connection.js
                </a>
              
                
                <a class="source" href="../Connection/ConnectionActions.html">
                  src/components/Connection/ConnectionActions.js
                </a>
              
                
                <a class="source" href="../Connection/ConnectionReducer.html">
                  src/components/Connection/ConnectionReducer.js
                </a>
              
                
                <a class="source" href="../Connection/messages.html">
                  src/components/Connection/messages.js
                </a>
              
                
                <a class="source" href="../Intro/Intro.html">
                  src/components/Intro/Intro.js
                </a>
              
                
                <a class="source" href="../Intro/IntroActions.html">
                  src/components/Intro/IntroActions.js
                </a>
              
                
                <a class="source" href="../Particles/Particles.html">
                  src/components/Particles/Particles.js
                </a>
              
                
                <a class="source" href="../Pulse/Pulse.html">
                  src/components/Pulse/Pulse.js
                </a>
              
                
                <a class="source" href="../Router/Router.html">
                  src/components/Router/Router.js
                </a>
              
                
                <a class="source" href="CatSound.html">
                  src/components/Sound/CatSound.js
                </a>
              
                
                <a class="source" href="Mute.html">
                  src/components/Sound/Mute.js
                </a>
              
                
                <a class="source" href="Sound.html">
                  src/components/Sound/Sound.js
                </a>
              
                
                <a class="source" href="TurningSound.html">
                  src/components/Sound/TurningSound.js
                </a>
              
                
                <a class="source" href="../Spinner/Spinner.html">
                  src/components/Spinner/Spinner.js
                </a>
              
                
                <a class="source" href="../Typing/Typing.html">
                  src/components/Typing/Typing.js
                </a>
              
                
                <a class="source" href="../../configureStore.html">
                  src/configureStore.js
                </a>
              
                
                <a class="source" href="../../index.html">
                  src/index.js
                </a>
              
                
                <a class="source" href="../../serviceWorker.html">
                  src/serviceWorker.js
                </a>
              
                
                <a class="source" href="../../setupTests.html">
                  src/setupTests.js
                </a>
              
            </div>
          </div>
        </li>
      </ul>
    
    <ul class="sections">
        
          <li id="title">
              <div class="annotation">
                  <h1>CatSound.js</h1>
              </div>
          </li>
        
        
        
        <li id="section-1">
            <div class="annotation">
              
              <div class="pilwrap ">
                <a class="pilcrow" href="#section-1">&#182;</a>
              </div>
              <p>Cat sound</p>

            </div>
            
            <div class="content"><div class='highlight'><pre>
<span class="hljs-keyword">import</span> React, { memo, forwardRef } <span class="hljs-keyword">from</span> <span class="hljs-string">'react'</span>;

<span class="hljs-keyword">export</span> <span class="hljs-keyword">const</span> CAT_SOUND = <span class="hljs-string">`<span class="hljs-subst">${process.env.PUBLIC_URL}</span>/catSound.mp3`</span>;

<span class="hljs-keyword">export</span> <span class="hljs-keyword">const</span> CatSound = forwardRef(<span class="hljs-function">(<span class="hljs-params">props, ref</span>) =&gt;</span> {
  <span class="hljs-keyword">return</span> (
    <span class="xml"><span class="hljs-tag">&lt;<span class="hljs-name">audio</span> <span class="hljs-attr">ref</span>=<span class="hljs-string">{ref}</span> <span class="hljs-attr">data-testid</span>=<span class="hljs-string">"cat-audio"</span>&gt;</span>
      <span class="hljs-tag">&lt;<span class="hljs-name">source</span> <span class="hljs-attr">src</span>=<span class="hljs-string">{CAT_SOUND}</span> <span class="hljs-attr">preload</span>=<span class="hljs-string">"auto"</span> /&gt;</span>
    <span class="hljs-tag">&lt;/<span class="hljs-name">audio</span>&gt;</span>
  );
});

export default memo(CatSound);

</span></pre></div></div>
            
        </li>
        
    </ul>
  </div>
</body>
</html>
