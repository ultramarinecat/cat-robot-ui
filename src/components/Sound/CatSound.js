//
// Cat sound
//

import React, { memo, forwardRef } from 'react';

export const CAT_SOUND = `${process.env.PUBLIC_URL}/catSound.mp3`;

export const CatSound = forwardRef((props, ref) => {
  return (
    <audio ref={ref} data-testid="cat-audio">
      <source src={CAT_SOUND} preload="auto" />
    </audio>
  );
});

export default memo(CatSound);
