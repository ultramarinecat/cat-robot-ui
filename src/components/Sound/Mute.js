//
// Mute button
// Updates icon based on muted state
//

import React, { memo } from 'react';
import PropTypes from 'prop-types';

import styles from './Mute.module.scss';

const MUTE_ICON = 'volume_down';
const UNMUTE_ICON = 'volume_mute';

export function Mute({ isMuted, onClick }) {
  const muteButtonIcon = isMuted ? UNMUTE_ICON : MUTE_ICON;

  return (
    <span onClick={onClick} data-testid="mute">
      <i
        className={`${styles.main} material-icons`}
        data-testid={isMuted ? 'unmute-icon' : 'mute-icon'}
      >
        {muteButtonIcon}
      </i>
    </span>
  );
}

Mute.propTypes = {
  isMuted: PropTypes.bool,
  onClick: PropTypes.func.isRequired
};

export default memo(Mute);
