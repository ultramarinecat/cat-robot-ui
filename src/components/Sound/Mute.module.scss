@charset 'utf-8';

////
//// Mute button styles
////

@import 'base.scss';

.main {
  font-size: rem(18);
  position: absolute;
  right: rem(5);
  bottom: rem(-35);
}
