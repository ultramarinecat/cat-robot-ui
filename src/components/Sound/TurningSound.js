//
// Turn request sound
//

import React, { memo, forwardRef } from 'react';

const TURNING_SOUND = `${process.env.PUBLIC_URL}/turningSound.mp3`;

export const TurningSound = forwardRef((props, ref) => {
  return (
    <audio ref={ref} data-testid="turning-audio">
      <source src={TURNING_SOUND} preload="auto" />
    </audio>
  );
});

export default memo(TurningSound);
