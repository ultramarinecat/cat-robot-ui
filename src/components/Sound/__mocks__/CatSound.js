import React, { forwardRef } from 'react';

export const playCat = jest.fn();
const pause = jest.fn();

const CatSound = forwardRef((props, ref) => {
  ref.current = {
    play: playCat,
    pause
  };

  return <div data-testid="cat-audio" />;
});

export default CatSound;
