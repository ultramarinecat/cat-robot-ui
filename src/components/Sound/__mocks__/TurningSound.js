import React, { forwardRef } from 'react';

export const playTurning = jest.fn();
const pause = jest.fn();

const TurningSound = forwardRef((props, ref) => {
  ref.current = {
    play: playTurning,
    pause
  };

  return <div data-testid="turning-audio" />;
});

export default TurningSound;
