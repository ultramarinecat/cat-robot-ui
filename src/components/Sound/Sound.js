//
// Sound component
// Displays mute button, plays cat and turning sounds
//

import React, { memo, useState, useEffect, useCallback, useRef } from 'react';
import PropTypes from 'prop-types';

import CatSound from './CatSound.js';
import TurningSound from './TurningSound.js';
import Mute from './Mute.js';

export function Sound({ playCatSound, playTurningSound, updateSound }) {
  // update even if sounds haven't changed
  const [muted, setMuted] = useState(true);

  const catSound = useRef();
  const turningSound = useRef();
  const isInitialClick = useRef(true);

  useEffect(() => {
    if (muted || (!playCatSound && !playTurningSound)) {
      return;
    }

    playSound(playCatSound ? catSound.current : turningSound.current);
  }, [playCatSound, playTurningSound, updateSound]);

  const handleClick = useCallback(() => {
    setMuted(isMuted => {
      // toggle mute setting, play turning sound if unmuting
      if (isMuted) {
        playSound(turningSound.current);
      }
      return !isMuted;
    });

    // user interaction required to play audio
    if (isInitialClick.current) {
      isInitialClick.current = false;
      playSound(catSound.current, true);
    }
  });

  return (
    <>
      <CatSound ref={catSound} />
      <TurningSound ref={turningSound} />

      <Mute isMuted={muted} onClick={handleClick} />
    </>
  );
}

function playSound(sound, muted = false) {
  sound.pause();
  sound.currentTime = 0;
  sound.muted = muted;

  sound.play();
}

Sound.propTypes = {
  playCatSound: PropTypes.bool,
  playTurningSound: PropTypes.bool,
  updateSound: PropTypes.bool
};

export default memo(Sound);
