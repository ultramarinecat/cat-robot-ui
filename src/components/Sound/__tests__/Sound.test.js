import React from 'react';
import { render, fireEvent } from 'react-testing-library';

import { Sound } from '../Sound.js';
import { playCat } from '../CatSound.js';
import { playTurning } from '../TurningSound.js';

jest.mock('../CatSound');
jest.mock('../TurningSound');

const MUTE_BUTTON = 'mute';
const CAT_AUDIO = 'cat-audio';
const TURNING_AUDIO = 'turning-audio';

describe('Sound', () => {
  afterEach(() => {
    playCat.mockClear();
    playTurning.mockClear();
  });

  it('should play cat and turning sounds and display mute button', () => {
    const { getByTestId, container } = render(<Sound />);

    expect(getByTestId(CAT_AUDIO)).toBeInTheDocument();
    expect(getByTestId(TURNING_AUDIO)).toBeInTheDocument();
    expect(getByTestId(MUTE_BUTTON)).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should initially be muted', () => {
    render(<Sound playCatSound={true} />);
    expect(playCat).not.toHaveBeenCalled();
  });

  it('should play sounds after unmute button click', () => {
    const { getByTestId, rerender } = render(<Sound playCatSound={true} />);
    expect(playCat).not.toHaveBeenCalled();

    fireEvent.click(getByTestId(MUTE_BUTTON));

    rerender(<Sound playCatSound={true} />);
    expect(playCat).toHaveBeenCalled();
  });

  it('should not play sounds after mute button click', () => {
    const { getByTestId, rerender } = render(<Sound playCatSound={true} />);
    fireEvent.click(getByTestId(MUTE_BUTTON));

    rerender(<Sound playCatSound={true} />);
    expect(playCat).toHaveBeenCalled();

    playCat.mockClear();
    fireEvent.click(getByTestId(MUTE_BUTTON));

    rerender(<Sound playCatSound={true} />);
    expect(playCat).not.toHaveBeenCalled();
  });

  it('should play turning sound when unmuted', () => {
    const { getByTestId } = render(<Sound />);

    expect(playTurning).not.toHaveBeenCalled();

    fireEvent.click(getByTestId(MUTE_BUTTON));
    expect(playTurning).toHaveBeenCalled();

    playTurning.mockClear();

    fireEvent.click(getByTestId(MUTE_BUTTON));
    expect(playTurning).not.toHaveBeenCalled();

    fireEvent.click(getByTestId(MUTE_BUTTON));
    expect(playTurning).toHaveBeenCalled();
  });

  it('should play cat sound', () => {
    const { getByTestId, rerender } = render(<Sound playCatSound={true} />);
    fireEvent.click(getByTestId(MUTE_BUTTON));

    rerender(<Sound playCatSound={true} />);
    expect(playCat).toHaveBeenCalled();
  });

  it('should play turning sound', () => {
    const { getByTestId, rerender } = render(<Sound playTurningSound={true} />);
    fireEvent.click(getByTestId(MUTE_BUTTON));

    rerender(<Sound playTurningSound={true} />);
    expect(playTurning).toHaveBeenCalled();
  });
});
