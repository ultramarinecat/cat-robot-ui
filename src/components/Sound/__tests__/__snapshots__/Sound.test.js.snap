// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sound should play cat and turning sounds and display mute button 1`] = `
<div
  data-testid="cat-audio"
/>
`;
